### ISSUE TITLE
contains fun and builders for Array

### ISSUE THREAD

**Issue by robstoll** [robstoll/atrium#11]
So far one has to use `asIterable` when it comes to `Array`. 
Add corresponding extension functions for `Array<out T>` which delegate to the `Iterable<T>` by using `asIterable` internally.
I think it is an easy starting point to see how Atrium supports multiple APIs.
The tasks of this issue are:
- copy `iterableAssertions.kt` in atrium-api-cc-en_UK and paste it as `arrayAssertions.kt`
  - replace `Iterable<T>` with `Array<out T>`
  - adjust KDoc if necessary
- repeat the same for atrium-api-cc-de_CH
- repeat the same for atrium-api-cc-infix-en_UK

**Comment by philipsdoctor** [comment]
I took a swing at this for an hour.  The problems start with "Upper bound of a type parameter cannot be an array"
```
val <E, T : Array<out E>> Assert<T>.contains: IterableContainsBuilder<E, T, IterableContainsNoOpSearchBehaviour>
    get() = _containsBuilder(this)
```

It spirals from there, I tried a few times with eliminating the bounding of T as something like:

```
val <T> Assert<Array<T>>.contains: ArrayContainsBuilder<IterableContainsNoOpSearchBehaviour>
    get() = _containsBuilder(this)

class ArrayContainsBuilder<out T, out S : IterableContains.SearchBehaviour>(
    plant: AssertionPlant<Array<out T>>, searchBehaviour: S
) : ContainsBuilder<Array<out T>, S>(plant, searchBehaviour)
```
This ended up breaking down when I tried writing a test and got a slew of type inference failed, I think I literally need to

```
val Assert<FloatArray>.contains: FloatArrayContainsBuilder<IterableContainsNoOpSearchBehaviour>
    get() = _containsBuilder(this)

```
And repeat that exercise for all of the assertions and builders and for every array type.  Feels a bit thwarting.

I feel like perhaps 
```
val <E> Assert<Array<E>>.contains: IterableContainsBuilder<E, Iterable<E>, IterableContainsNoOpSearchBehaviour>
    get() = _containsBuilder(returnValueOf(this.subject::asIterable))
```
or something similar is in order, just make the type transformation easy for the consumer?

Maybe I'm just missing something basic, the kotlin docs just tell me to read Effective Java in order to understand this, but the link provided is "how to buy"

https://kotlinlang.org/docs/reference/generics.html

I think the short of it is that arrays aren't using generics, so we can't write a generic solution.  I'm happy to be wrong on this one. :)

**Comment by robstoll** [comment]
First of all, thank you very much for spending time to improve atrium :+1: 

> I took a swing at this for an hour. The problems start with "Upper bound of a type parameter cannot be an array"
Interesting, I wasn't aware of that one, but I think your next step was good. 

I see now, that I should have give a further hint about `Array<out T>` and the implementation for primitive types. `FloatArray` isn't a subtype of `Array<out Float>` or `Array<Float>`, it is a completely different type, which mimicks `float[]`. At first I included those types in the Task but thought it is too repetitive to add assertion functions for those types as well and also too much would be required for a `help wanted` task and left out the comment at all. Well well... I can improve there. 

To sum up, I guess we are almost there and there is not much left to make it work. Would be great if you can write a pull request nonetheless (it does not matter if it has compilation errors or other problems). I have an easier starting point from there to resolve the remaining problems.

> I feel like perhaps ... or something similar is in order
In terms of maintenance I fully agree. All the additional functions just for leaving out `returnValueOf(subject::asIterable)` but I thought it would be worth that users of Atrium do not have to repeat that over and over again. I will decide after I saw your pull request, maybe it is worth the hassle.

**Comment by robstoll** [comment]
@philipsdoctor I came to the conclusion that you are right. It is not worth copying functions, it would be a maintenance nightmare. Atrium 0.7.0 will provide an `asIterable()` function for Arrays.
